conda/models/version.py

```diff
@@ -519,8 +519,14 @@ def get_matcher(self, vspec):
                 elif operator_str == "!=":
                     vo_str = vo_str[:-2]
                     operator_str = "!=startswith"
-                else:
+                elif operator_str == "~=":
                     raise InvalidVersionSpec(vspec_str, "invalid operator with '.*'")
+                else:
+                    log.warning("Using .* with relational operator is superfluous and deprecated "
+                                "and will be removed in a future version of conda. Your spec was "
+                                "{}, but conda is ignoring the .* and treating it as {}"
+                                .format(vo_str, vo_str[:-2]))
```

Comment:
Can it really be remove in a future version? Packages are out there with this versionspec and they will always be out there.

For me a critical benefit of using `conda` is reproducibility which means I can use `conda` to recreate an exactly specified environment and get the exact same analytics results back (possibly modulo floating point handling differences)

If new `conda` can't install historical packages/environments that would break the reproducibility/auditability feature which is crucial to our usage of it as a package manager.

Reply:
Yes, things can break. Nothing is permanent. People can use old versions of conda, or they can hotfix repodata. We won't do these things lightly, but they need to be possible.

This particular change is really not too important to split hairs over, and I'm honestly happy to ignore this as cruft, but there's a lot of old stuff in old packages that genuinely breaks things. If we are bound completely by the past, conda will not be able to move forward. There are sane limits to backwards compatibility. Docker containers and/or VMs should also be a part of your long term reproducibility strategy. Conda's primary concern is keeping things working correctly, not necessarily keeping old things working. Where those are at odds, old stuff will break.

Reply:
I'm sympathetic to need for progress and improvement in the code base and especially to the fact that historical packaging problems can blow things up through not fault of conda's (I've been bitten by that before). I've actually wondered if some sort of channel sharding wouldn't help alleviate these problems - i.e. move all packages older than X to a `historical` channel.

In any case, this particular bit of cruft seems pretty harmless to keep around, compared to the breakage it can induce.

I'd recommend keeping the warning (so people eventually fix their packages) but allow the conda install to proceed... pretty much indefinitely. What I would also do though is to disallow the incorrect versionspec in conda-build so that no new packages can make the same mistake.

Reply:
Yep, those are all good suggestions. I agree with all of them and hopefully we can work towards them.